### PR TITLE
Docs: Sort README by cost-benefit

### PR DESCRIPTION
**Pull request by brettz9**
Although I was hoping for PR #4555 to provide a table format for the Rules README which could be sorted along different dimensions, putting this into such a large table was dismissed as being too drastic a change.

So, I'm making a new PR, also with very many changes, but preserving the existing (non-tabular) format.

My own interest for using ESLint is to learn first about those rules which can fix my errors, with as little cost as possible, and slowly proceed to best practices which may come at a cost, and finally to purely stylistic issues, so that is the basis for this reorganization.

While it can be useful to know the environment of applicability (and I tried to preserve that information within sections), I thought that it is better to have one single criterion of organization (cost-benefit), and then denote the environment within the different cost-benefit sections.

Given the large number of changes, I suggest skipping the diffs and just looking at the new file as a whole.

### CODE REVIEW

**Comment by brettz9** [comment]
I suppose the docs tag is no longer valid here, given the fact that my suggested changes in "recommended" rules, required a couple of code changes as well, but I'm not sure which tag to use, and I figure this PR may either be rejected or require a bit of consideration first anyways.

File: conf/eslint.json [hunk 1]
```diff
@@ -88,10 +88,10 @@
         "no-return-assign": "off",
         "no-script-url": "off",
         "no-self-assign": "error",
-        "no-self-compare": "off",
+        "no-self-compare": "error",
```
Comment (nzakas) [inline conf/eslint.json]: Please revert all these changes, this affects everyone using `eslint:recommended`. If you're proposing to enable these rules for just the ESLint project, please make changes in `packages/eslint-config-eslint/default.yml`.

File: docs/rules/README.md [hunk 2]
```diff
@@ -1,151 +1,318 @@
...
 
-Some rules are fixable using the `--fix` command line flag. Those rules are marked as "(fixable)" below. Currently these fixes are mostly limited only to whitespace fixes.
+Rules in ESLint are divided into several categories to help you better
+understand their value, with the rules being introduced below roughly
+according to a descending benefit-cost value, and subgrouped according
+to whether the ratio is due more to benefit or cost, and according to
+whether the rule applies only in specific environments.
 
-## Possible Errors
+All rules are disabled by default.
```
Comment (nzakas) [inline docs/rules/README.md]: :-1: Most people are coming to this page for the list of rules and skip over the text at the beginning. We need the text to be as short as possible and put the focus on the rules list. I think this adds way too much text up front and the more verbose grouping is a bit too much.

File: conf/eslint.json (hunk 1 above)
Comment (brettz9) [inline conf/eslint.json]: Ok, but my intent was to propose enabling them for everyone using `eslint:recommended`. :) Is it too late to change these and/or do you not agree they are fairly serious enough (or easy enough to workaround) so as to merit error messages (I could see the `blocked-scope-var` being a little more arguable than the others)?

File: docs/rules/README.md (hunk 2 above)
Comment (BYK) [inline docs/rules/README.md]: :+1: to @nzakas

File: conf/eslint.json (hunk 1 above)
Comment (ilyavolodin) [inline conf/eslint.json]: We can only change those during major releases. So the next change will be during 3.0 release.

**Comment by brettz9** [comment]
I hope I can keep my own fork's readme branch up to date with your own readme changes.

**Pull request closed without merging**